packages/@react-aria/landmark/src/useLandmark.ts

```diff
@@ -0,0 +1,208 @@
+  let walker = document.createTreeWalker(
+    root,
+    NodeFilter.SHOW_ELEMENT,
+    {
+      acceptNode(node) {
+        // Not sure we need this, can landmarks be nested?
+        // Skip nodes inside the starting node.
+        // if (opts?.from?.contains(node)) {
+        //   return NodeFilter.FILTER_REJECT;
+        // }
+
+        if ((node as HTMLElement).matches(selector)
+          && isElementVisible(node as HTMLElement)
+          && opts?.registeredLandmarks.includes(node)) {
+          return NodeFilter.FILTER_ACCEPT;
+        }
+
+        return NodeFilter.FILTER_SKIP;
+      }
+    }
+  );
+
+  if (opts?.from) {
+    walker.currentNode = opts.from;
+  }
+
+  return walker;
+}
+
+function treeDistance(child: HTMLElement, target: HTMLElement): number {
+  let distance = 0;
+  let node = child;
+  while (target && node.parentElement && node !== target) {
+    node = node.parentElement;
+    distance++;
+  }
+  return distance;
+}
+
+class LandmarkManager {
+  private landmarks = new Map<MutableRefObject<HTMLElement>, HTMLElement>();
+  private static instance: LandmarkManager;
+
+  private constructor() {}
+
+  public static getInstance(): LandmarkManager {
+    if (!LandmarkManager.instance) {
+      LandmarkManager.instance = new LandmarkManager();
+    }
+
+    return LandmarkManager.instance;
+  }
+
+  public addLandmark(ref: MutableRefObject<HTMLElement>) {
+    // could change to a map to track roles, then we could warn if someone provides two of the same role but no labels for them
+    if (!this.landmarks.has(ref)) {
+      this.landmarks.set(ref, null);
+    }
+  }
+
+  public removeLandmark(ref: MutableRefObject<HTMLElement>) {
+    if (this.landmarks.has(ref)) {
+      this.landmarks.delete(ref);
+    }
+  }
+
+  // Gets the landmark that is the closest parent in the DOM to the child
+  // Useful for nested Landmarks
+  private closestLandmark(child: HTMLElement) {
+    return Array.from(this.landmarks.keys())
+      .filter(landmark => (landmark?.current as HTMLElement).contains(child))
+      .sort((landmarkA, landmarkB) =>
+        treeDistance(child, landmarkA?.current) - treeDistance(child, landmarkB?.current)
+      )[0];
+  }
+
+  public getNextLandmark(walker, {backward}) {
+    let nextLandmark;
+    if (backward) {
+      nextLandmark = walker.previousNode();
+      if (!nextLandmark) {
+        walker.currentNode = document.body;
+        // in nested cases, this won't return any of the nested, so traverse down the tree until there is not longer a lastChild
+        nextLandmark = walker.lastChild();
+        let temp = nextLandmark;
+        while (temp) {
+          nextLandmark = temp;
+          walker.currentNode = temp;
+          temp = walker.lastChild();
+        }
+      }
+    } else {
+      nextLandmark = walker.nextNode();
+      if (!nextLandmark) {
+        walker.currentNode = document.body;
+        nextLandmark = walker.firstChild();
+      }
+    }
+    return nextLandmark;
```

Comment:
I think you could avoid walking the entire DOM tree if you kept the landmarks sorted in document order. When inserting, you could compare the position of the new node vs the other nodes with the compareDocumentPosition method with binary search: https://developer.mozilla.org/en-US/docs/Web/API/Node/compareDocumentPosition. Then when pressing F6, you only need to find the current landmark, and go to the next/previous one vs search the whole DOM potentially.